### PR TITLE
Migrate to "node2" debug engine

### PR DESCRIPTION
**Pull request by vladimir-kotikov**
Use "node2" debugging engine (https://github.com/Microsoft/vscode-node-debug2) rather than "node" for debugging react-native code. As mentioned in #377 previous approach to debugging will be deprecated in Node 8.

This should resolve a lot of debugging issues we've seen previously, such as #367 and #356

### CODE REVIEW

**Comment by vladimir-kotikov** [comment]
@MSLaguana, @roblourens I would really appreciate any feedback on this

File: src/debugger/forkedAppWorker.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import { Log } from "../common/log/log";
+import { LogLevel } from "../common/log/logHelper";
+import { ErrorHelper } from "../common/error/errorHelper";
+import { IDebuggeeWorker, RNAppMessage } from "./appWorker";
+
+// tslint:disable-next-line:align
+const WORKER_BOOTSTRAP = `
+// Hacks to avoid accessing unitialized variables
+var onmessage=null, self=global;
```
Comment (MSLaguana) [inline src/debugger/forkedAppWorker.ts]: Should you also reference `importScripts` here?

File: src/debugger/forkedAppWorker.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+// Avoid Node's GLOBAL deprecation warning
+global.GLOBAL=global;
+process.on("message", function(message){
+    if (onmessage) onmessage(message);
+});
+postMessage = function(message){
+    process.send(message);
+};
+importScripts = function(scriptUrl){
+    var scriptCode = require("fs").readFileSync(scriptUrl, "utf8");
```
Comment (MSLaguana) [inline src/debugger/forkedAppWorker.ts]: I'm cautious about these `require`s here. I know that react-native redefines global require, so I'm not sure whether that can impact this if `importScripts` is called more than once. Might be worth using an IIFE to take a reference to the original require function initially and use that when `importScripts` is called? Or even just cache both `fs` and `vm` immediately and use those in `importScripts`

Comment (vladimir-kotikov) [inline src/debugger/forkedAppWorker.ts]: Great idea! I totally missed that fact with `require` override by react

File: src/debugger/forkedAppWorker.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                `Spawned debuggee process with pid ${this.debuggeeProcess.pid} listening to ${port}`);
+            return port;
+        });
+    }
+
+    public postMessage(rnMessage: RNAppMessage): void {
+        // Before sending messages, make sure that the worker is loaded
+        this.workerLoaded.promise
+        .then(() => {
+            if (rnMessage.method !== "executeApplicationScript") return Q.resolve(rnMessage);
```
Comment (MSLaguana) [inline src/debugger/forkedAppWorker.ts]: Shouldn't other messages be passed on to the other process to be run in there?

Comment (MSLaguana) [inline src/debugger/forkedAppWorker.ts]: Oh, I see that is handled in the `done` coming up next.

File: src/debugger/forkedAppWorker.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+process.on("message", function(message){
+    if (onmessage) onmessage(message);
+});
+postMessage = function(message){
+    process.send(message);
+};
+importScripts = function(scriptUrl){
+    var scriptCode = require("fs").readFileSync(scriptUrl, "utf8");
+    require("vm").runInThisContext(scriptCode, { filename: scriptUrl });
+};`;
```
Comment (MSLaguana) [inline src/debugger/forkedAppWorker.ts]: Something else that I would recommend is adding a global `__debug__` object with a `require` function which caches the node `require`. This matches what we currently expose in the original sandboxing method, and allows things such as realm to provide shims for native functionality when necessary.

**Comment by roblourens** [comment]
LGTM. You may want to follow this issue - https://github.com/Microsoft/vscode/issues/19650 - as we're talking about how the two debug adapters will interact in the future. It probably won't mean anything for you, at least for February. There's a chance they could end up merged into a single extension or something, but probably not.

**Comment by MSLaguana** [comment]
On the whole this looks pretty good to me. Am I correct that the SingleLifetimeAppWorker is no longer used at all, and is replaced by the forking app worker?

**Comment by vladimir-kotikov** [comment]
Thanks for review, guys!

> Am I correct that the SingleLifetimeAppWorker is no longer used at all, and is replaced by the forking app worker?

That's correct. I initially thought that we might want to keep both them, just in case if we plan to support old "node" debugger, but this doesn't seem to be the case. Am i right? Do you want me to drop that old worker code?

File: src/debugger/appWorker.ts [hunk 5]
```diff
@@ -319,16 +315,24 @@ export class MultipleLifetimesAppWorker extends EventEmitter {
     }
 
     private onMessage(message: string) {
+        function killWorker(worker: IDebuggeeWorker) {
+            if (!worker) return;
+            worker.stop();
+            worker = null;
```
Comment (MSLaguana) [inline src/debugger/appWorker.ts]: This won't modify `this.worker` which I believe was your intent.

**Comment by MSLaguana** [comment]
Looks like one of our tests is failing now, but it seems to be specifically to do with the single lifetime worker which isn't as relevant any more. Could you please look into that and update or remove the test as appropriate?

**Comment by MSLaguana** [comment]
Also to answer your question about the single lifetime worker: I think it would be best to only have the one way of doing things, so I am inclined to remove the single lifetime worker and just focus on the forked worker.

**Comment by vladimir-kotikov** [comment]
@MSLaguana, addressed your comments and updated tests to actually test current worker implementation. Please take a look once more

File: src/test/debugger/appWorker.test.ts [hunk 6]
```diff
@@ -6,80 +6,68 @@ import * as WebSocket from "ws";
...
+                return workerWithScript(startScriptContents).start()
+                .then(() =>
+                    Q.delay(100))
+                .then(() =>
+                    assert(postReplyFunction.calledWithExactly(expectedMessageResult)));
             });
 
             test("should be able to import scripts", function() {
-                const scriptImportPath = "testScript.js";
+                // NOTE: we're not able to mock out
```
Comment (MSLaguana) [inline src/test/debugger/appWorker.test.ts]: Looks like this comment is unfinished?

File: src/test/debugger/appWorker.test.ts [hunk 7]
```diff
@@ -6,80 +6,68 @@ import * as WebSocket from "ws";
...
-                }));
-
-                return sandboxedWorker.start().then(() => {
+
+                return workerWithScript(startScriptContents).start().then(() => {
                     // We have not yet finished importing the script, we should not have posted a response yet
                     assert(postReplyFunction.notCalled, "postReplyFuncton called before scripts imported");
-                    scriptImportDeferred.resolve(void 0);
-                    return Q.delay(1);
+                    return Q.delay(100);
```
Comment (MSLaguana) [inline src/test/debugger/appWorker.test.ts]: What's the reasoning behind increasing this limit here? Before we used 1 just to make sure that it was asynchronous. Is the delay of 100ms actually required for processing to happen on time, are there potential race conditions here?

**Comment by MSLaguana** [comment]
Just a couple of questions, looking good.

File: src/test/debugger/appWorker.test.ts (hunk 6 above)
Comment (vladimir-kotikov) [inline src/test/debugger/appWorker.test.ts]: Yeah, there should be a description of reasoning why we have to use real script file to test `importScripts`. I'll update.

File: src/test/debugger/appWorker.test.ts (hunk 7 above)
Comment (vladimir-kotikov) [inline src/test/debugger/appWorker.test.ts]: I think that previously the timeout was required just to delay dispatching the current promise chain and let `scriptImportDeferred` chain to be dispatched first.

In this implementation the spawned process is a black box and completely async for us so we can't reliably determine when the script has been imported and executed, so the only way to test this IMO is to give some reasonable amount of time to worker to complete script's import before checking if it has been done.

As for race conditions - not sure where are they might happen. Every test spawns it's own worker, so they are isolated from each other.

**Comment by MSLaguana** [comment]
Looks great, thanks @vladimir-kotikov!

**Pull request merged**